sensing/image_diagnostics/src/image_diagnostics_node.cpp

```diff
@@ -0,0 +1,280 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "image_diagnostics/image_diagnostics_node.hpp"
+
+#include <std_msgs/msg/header.hpp>
+
+namespace image_diagnostics
+{
+using image_diagnostics::Image_State;
+ImageDiagNode::ImageDiagNode(const rclcpp::NodeOptions & node_options)
+: Node("image_diagnostics_node", node_options)
+{
+  image_sub_ = create_subscription<sensor_msgs::msg::Image>(
+    "input/raw_image", rclcpp::SensorDataQoS(),
+    std::bind(&ImageDiagNode::ImageChecker, this, std::placeholders::_1));
+  block_diag_image_pub_ =
+    image_transport::create_publisher(this, "image_diag/debug/diag_block_image");
+  dft_image_pub_ = image_transport::create_publisher(this, "image_diag/debug/dft_image");
+  gray_image_pub_ = image_transport::create_publisher(this, "image_diag/debug/gray_image");
+  raw_image_pub_ = image_transport::create_publisher(this, "image_diag/raw_image");
+  frequency_intensity_pub1_ = create_publisher<tier4_debug_msgs::msg::Float32Stamped>(
+    "image_diag/frequency_intensity_ratio", rclcpp::SensorDataQoS());
+
+  image_state_pub_ = create_publisher<tier4_debug_msgs::msg::Int32Stamped>(
+    "image_diag/image_state_diag", rclcpp::SensorDataQoS());
+
+  updater_.setHardwareID("Image_Diagnostics");
+  updater_.add(
+    std::string(this->get_namespace()) + ": Image_Diagnostics", this,
+    &ImageDiagNode::onImageDiagChecker);
+  updater_.setPeriod(0.1);
+}
+
+void ImageDiagNode::onImageDiagChecker(DiagnosticStatusWrapper & stat)
+{
+  stat.add("number dark blocks ", std::to_string(num_of_regions_dark));
+  stat.add("number blockaged blocks ", std::to_string(num_of_regions_blockage));
+  stat.add("number low visibility blocks ", std::to_string(num_of_regions_low_visibility));
+  stat.add("number backlight  blocks ", std::to_string(num_of_regions_backlight));
+
+  auto level = DiagnosticStatusWrapper::OK;
+  if (diagnostic_status_ < 0) {
+    level = DiagnosticStatusWrapper::STALE;
+  } else if (diagnostic_status_ == 1) {
+    level = DiagnosticStatusWrapper::WARN;
+  } else if (diagnostic_status_ == 2) {
+    level = DiagnosticStatusWrapper::ERROR;
+  } else {
+    level = DiagnosticStatusWrapper::OK;
+  }
+
+  std::string msg;
+  if (level == DiagnosticStatusWrapper::OK) {
+    msg = "OK";
+  } else if (level == DiagnosticStatusWrapper::WARN) {
+    msg = "WARNING: abnormal state in image diagnostics";
+  } else if (level == DiagnosticStatusWrapper::ERROR) {
+    msg = "ERROR: abnormal state in image diagnostics";
+  } else if (level == DiagnosticStatusWrapper::STALE) {
+    msg = "STALE";
+  }
+  stat.summary(level, msg);
+}
+
+void ImageDiagNode::ImageChecker(const sensor_msgs::msg::Image::ConstSharedPtr input_image_msg)
+{
+  cv::Mat img_gray;
+  cv::Mat img_gray_32b;
+  cv::Mat img_gray_blockage_bin;
+  cv::Mat tmp;
+
+  raw_image_pub_.publish(input_image_msg);
```

Comment:
[Q] Why it publishes original image? Isn't this a duplicate?

Reply:
I will remove it